### ISSUE TITLE
Array.removeAt

### ISSUE THREAD

**Issue by delahee** [HaxeFoundation/haxe#4238]
Hi,

Deepnight and I think haxe Array misses : 
```
function removeAt( pos : Int) : Void;
```
when there is only :
```
function splice( pos : Int, len : Int ) : Array<T>;
```
Splice is very problematique since int creates another array and totally trashes memory. On most platforms this function already exists so here is a big astraction leak that haxe should fill.

for example on cpp, ths function would not ungrow the memory zone but use a memmove to retrieve the overlappig pointer zone, thus providing a very fast imlementation.

So we propose we introduce this function in cross platform.

we can start with a naive implementation which is 

```
function removeAt( pos : Int) : Void {
arr.remove( arr[idx] );
}
```

For informations, array is very often the (very) worst offender in memory usage yet the best ergonomic alternative, so we think improving it would be great.

**Comment by nadako** [comment]
I remember @waneck wanted to analyze unused `splice` results and generate `spliceVoid` method call that doesn't create another array (that's for C#/Java aka gencommon), but `removeAt` looks like a cleaner solution. However as always we're stuck with native flash/js implementation and discussion of whether these additional methods should be available thru Dynamic and reflection (see #3776, #3323)

**Comment by Simn** [comment]
Modifying operations like this should be on Array itself so they can be properly optimized. That said it's still going to require some effort to implement this and then there are still going to be issues if it is used through Dynamic, e.g. in templates.

I'm inclined to think that optimizing splice would be the better approach overall. In the static analyzer it would be easy to detect that its return value is not being used and then replace it with something the target generators can optimize.

**Comment by delahee** [comment]
@simn I am not hot with the static analyzer option it feels too optionnal and avoiding this allocation should be mainstream. 

I didn't get that there was an issue with reflection, maybe its the good time to design a true cross platform solutions for extending Array ?

thx !

**Comment by ncannasse** [comment]
I would try to push too much index based add/remove on Array, since it's a O(n) operation that can be quite inefficient compared to using a Map for instance

**Comment by delahee** [comment]
@ncannasse just a comment...your poin is true only for big n, since it is cache aware and array allocation should be amortised, it is often faster to memmove than to use allocating + cache unaware tree for small n's. Just like the same wau insertion sort is faster than bubble sort for small n's. In any case if it has native alternative removeAt will always be faster than splice and remove :)... 

Anyway as a true side note, we would be happy enough if splice() could just not perform the allocation when possible, it would be sufficient for us.

thanks !

**Comment by ncannasse** [comment]
I think that Array.remove does such thing already (not on Flash/JS since it uses splice)

**Comment by deltaluca** [comment]
````removeAt```` could not be implemented as ````arr.remove( arr[idx] );```` as this would fail in general if the array have duplicates by equality, eg:

````
var xs = [0, 1, 0];
xs.remove(xs[2]);
trace(xs); // [1, 0]
````

Now, if you 'did' consider that to be fine (order is not important in the Array), then you may aswell write implement as:

````
function removeAt(ind) {
    if (ind != length) this[ind] = this.pop();
}
````

Not that such a functino should be really named 'removeAt' anymore.

**Comment by waneck** [comment]
I agree that going through the static analyzer and optimizing it when we can is probably the best solution. It's not optional since it should be included by default on the next version of Haxe

**Comment by MSGhero** [comment]
Flash is getting `removeAt` and `insertAt`, not sure if it's a wrapper for splice or a whole new function: https://twitter.com/liquidate/status/621827033757126656

**Comment by delahee** [comment]
Ahah that is a good lesson. We should stop being so conservative about such
things. :)

2015-07-17 3:34 GMT+02:00 Nick <notifications@github.com>:

> Flash is getting removeAt and insertAt, not sure if it's a wrapper for
> splice or a whole new function:
> https://twitter.com/liquidate/status/621827033757126656
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/issues/4238#issuecomment-122144343>
> .
>



-- 
David Elahee

**Comment by Justinfront** [comment]
My day job is with as3 and there are a few nice things that surpise me, but Vectors are pretty crap, I just think Adobe should sit down with Haxe community and work something out. Haxe is pretty much as4+.

**Comment by delahee** [comment]
Ok now adobe made the thing, I'll hack it into my haxe std libs and send a pull request.

**Comment by MSGhero** [comment]
FP 19 was released earlier this week. Yay!

**Comment by delahee** [comment]
Yeah, definitive end of array removal that causes allocations !

**Comment by delahee** [comment]
ok work done, i'll list pull request here.

**Comment by delahee** [comment]
Pull requests done with tests, waiting for feedbacks :) thx !

**Comment by delahee** [comment]
Ok I think I know why the test fails, I ll put barriers in the test..

@andyli for the test to pass on cpp, hxcpp should be in sync with the PR...How should I proceed ?

**Comment by andyli** [comment]
If the PRs to both sides depend on each other, than there is no perfect way. My suggestion is to just ignore the build failure and check manually that the PRs are correct. Optionally, once both PRs are merged, restart the failed build to let it pick up the new commits from another side.

**Comment by delahee** [comment]
@andyli I think https://github.com/HaxeFoundation/hxcpp/pull/328 is pretty much standalone, when it passes, https://github.com/HaxeFoundation/haxe/pull/4633 will be ok with one fix.I'll wait for you or @hughsando to check and accept #328 then we can move on, in the mean time, i'll fix the utest tomorrow. thx.

**Comment by hughsando** [comment]
So I have implemented this, but if your goal it to reduce allocations, then you should not have a return value, or allow it to be unspecified in the case of the index being out of range.  Requiring a null return forces boxing of the result, even if it is ignored.
I would also like to suggest a "count" parameter, but then you might end up with the "is it count, or is it last" debacle we have with slice/splice.

**Comment by delahee** [comment]
Thanks you ! What is allocated then a Null<null> ?

I would vote we thus remove the return parameters.

What do you think ?

**Comment by Justinfront** [comment]
function removeAt( 5, ?returnValue: Boolean = true ); 
I feel it's important functionally and practically to return a null or the item removed by default, but it would be acceptable for it to return Void if returnValue was set to true.

**Comment by delahee** [comment]
This would be strange to have return on demand, feels like splice... this function was designed for allocation performance (otherwise splice is already here) so I think really prefer to remove any return so that we can keep best perfs...

**Comment by Justinfront** [comment]
I think for haxe to implement a method with a less functional approach and inconsistantly with as3 purely for performance seems to be against Haxe principles such a function should be in an external haxelib only, I don't think it should be added to the std if it does not implement a return.
http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/Array.html#removeAt%28%29

**Comment by ncannasse** [comment]
I agree we could remove the return.

**Comment by Justinfront** [comment]
well so be it, but I expect it will cause some future confusion for haxe libraries used by as3.

**Comment by delahee** [comment]
We are aware of that but haxe have to choose what is good on most platforms, since we have splice and removeAt's purpose is performance, we have to stick to our goals. If confusion there is we 'll explain in the function's doc.

**Comment by hughsando** [comment]
I think we can have the return value - as long as it is target dependent
when the index is out-of-range for numeric types.
I don't think this is going to hurt the usefulness since the only real case
I can think of where you might want null would be:

while( (result= array.removeAt(0) != null )
   result.destroy();

Hugh

**Comment by delahee** [comment]
@hughsando then I propose we simply return a boolean if something was effectively removed ?
I'd like to squeeze this one these days.

**Comment by delahee** [comment]
For the record, in a IRc chat, @Simn proposed to think about something with culling call site allocations. Let's see what comes around.

**Comment by Simn** [comment]
@hughsando: Do you think it would make sense to have non-allocating variants of some other Array methods which can be used for block-level calls? For instance, if we have `{ a.pop(); }` on an `Array<Int>` we don't have to deal with the `Null<T>` return because it doesn't matter.

**Comment by hughsando** [comment]
I have been think about this too.  Particularly splice, which creates a
whole array, that then gets discarded.
I would love a variant of map.get that does not need an allocation too.  My
recent work on the CPP ast may be able to detect when this is assigned to,
say, an Int and call "getInt" instead.  But it would be easier if it was
part of the Api.

**Comment by ncannasse** [comment]
@hughsando would it not be more simple to change the way you return Null values in general? (using a struct on stack instead of heap ptr)

**Comment by nadako** [comment]
btw, c# target uses `struct Null<T>` so it's stack allocated

**Comment by hughsando** [comment]
Yes, that is possible.  So is a Array of Null<Float> an array of structures of 12 bytes?
I have been thinking of using a Variant structure, like
```
string Variant
{
   int type;
   union
   {
        int ival;
        double dval;
       String sval;
       Dynamic dynamicVal;
   };
};
```
 for dynamic function calling and some function returns.  This would be 12/16 byes of 32/64 bit systems.

This is big change, and it is hard to see how to do it in such a was as to keep hxcpp compatible with earlier versions of haxe - although it should be possible.

**Comment by hughsando** [comment]
As for easier, I'm hoping using getInt should be very easy.

**Comment by ncannasse** [comment]
One quick hack would be that if there's a method of the same name but with a prefix + type suffix, you can call it instead.

For instance `.get()` will be translated to `.__hx__get_Int()` for instance
